File: coding/src/edu/algos/li20_dp/_Grid_TwoStartNMultiEnd_L1463_CherryPickupII.java
package edu.algos.li20_dp;

public class _Grid_TwoStartNMultiEnd_L1463_CherryPickupII {
    public static void main(String args[]){
        _Grid_TwoStartNMultiEnd_L1463_CherryPickupII obj = new _Grid_TwoStartNMultiEnd_L1463_CherryPickupII();

        int[][] grid = new int[][]{
            {3,1,1},
            {2,5,1},
            {1,5,5},
            {2,1,1}};
        System.out.println(obj.cherryPickup(grid));
        System.out.println(obj.cherryPickup2(grid));
    }

    // Problem: https://leetcode.com/problems/cherry-pickup-ii/
    // Idea: Idea is same like fixed start to variable ends. But there are 2 fixed positions and when they overlap, those cells should be counted only once.
    // Hence, we run both robots in the same recursion to eliminate the cases of overlapping. This leads to 3D dp.
    // Time: O(m*n*n*9)
    // Space: O(m*n*n)
    public int cherryPickup(int[][] grid) {
        int m=grid.length, n=grid[0].length;
        int[][][] cost = new int[m][n][n];
        for(int j1=0; j1<n; j1++){
            for(int j2=0; j2<n; j2++){
                if(j1==j2){
                    cost[m-1][j1][j2] = grid[m-1][j1];      // i=m-1
                }else{
                    cost[m-1][j1][j2] = grid[m-1][j1]+grid[m-1][j2];
                }
            }
        }

        for(int i=m-2; i>=0; i--){
            for(int j1=0; j1<n; j1++){
                for(int j2=0; j2<n; j2++){
                    int max = (int)-1e8;
                    for(int dj1=-1; dj1<=1; dj1++){
                        for(int dj2=-1; dj2<=1; dj2++){
                            int value = 0;
                            if(j1==j2){
                                value = grid[i][j1];
                            }else{
                                value = grid[i][j1]+grid[i][j2];
                            }
                            if(j1+dj1>=0 && j1+dj1<n && j2+dj2>=0 && j2+dj2<n)
                                value += cost[i+1][j1+dj1][j2+dj2];
                            else
                                value += (int)-1e8; 
                            max = Math.max(max, value);
                        }
                    }
                    cost[i][j1][j2] = max;
                }
            }
        }

        return cost[0][0][n-1];
    }

    public int cherryPickup2(int[][] grid) {
        int m=grid.length, n=grid[0].length;
        return cherryPickup2(grid, 0, 0, n-1, m, n);
    }

    private int cherryPickup2(int[][] grid, int i, int j1, int j2, int m, int n) {
        if(j1<0 || j2<0 || j1>=n || j2>=n)
            return (int)-1e8;
        if(i==m-1){
            if(j1==j2){
                return grid[i][j1];
            }else{
                return grid[i][j1]+grid[i][j2];
            }
        }

        int max = (int)-1e8;
        for(int dj1=-1; dj1<=1; dj1++){
            for(int dj2=-1; dj2<=1; dj2++){
                int value = 0;
                if(j1==j2){
                    value = grid[i][j1];
                }else{
                    value = (grid[i][j1]+grid[i][j2]);
                }
                value += cherryPickup2(grid, i+1, j1+dj1, j2+dj2, m, n);
                max = Math.max(max, value);
            }
        }
        return max;
    }
}

File: dsa/concepts/15_dp/python/grid_two_start_nmulti_end_l1463_cherry_pickup_ii.py
# Problem: https://leetcode.com/problems/cherry-pickup-ii/
# Idea: Idea is same like fixed start to variable ends. But there are 2 fixed positions and when they overlap, those cells should be counted only once.
# Hence, we run both robots in the same recursion to eliminate the cases of overlapping. This leads to 3D dp.
# Time: O(m*n*n*9)
# Space: O(m*n*n)
def cherryPickup(grid):
    m, n = len(grid), len(grid[0])

    cost = [[[0] * n for _ in range(n)] for _ in range(m)]

    for j1 in range(n):
        for j2 in range(n):
            if j1 == j2:
                cost[m - 1][j1][j2] = grid[m - 1][j1]      # i=m-1
            else:
                cost[m - 1][j1][j2] = grid[m - 1][j1] + grid[m - 1][j2]

    for i in range(m - 2, -1, -1):
        for j1 in range(n):
            for j2 in range(n):
                maxVal = int(-1e8)
                for dj1 in range(-1, 2):
                    for dj2 in range(-1, 2):
                        value = 0
                        if j1 == j2:
                            value = grid[i][j1]
                        else:
                            value = grid[i][j1] + grid[i][j2]

                        nj1, nj2 = j1 + dj1, j2 + dj2
                        if 0 <= nj1 < n and 0 <= nj2 < n:
                            value += cost[i + 1][nj1][nj2]
                        else:
                            value += int(-1e8)

                        maxVal = max(maxVal, value)

                cost[i][j1][j2] = maxVal

    return cost[0][0][n - 1]


def cherryPickup2(grid):
    m, n = len(grid), len(grid[0])
    return cherryPickup2_dfs(grid, 0, 0, n - 1, m, n)


def cherryPickup2_dfs(grid, i, j1, j2, m, n):
    if j1 < 0 or j2 < 0 or j1 >= n or j2 >= n:
        return int(-1e8)

    if i == m - 1:
        if j1 == j2:
            return grid[m - 1][j1]
        else:
            return grid[m - 1][j1] + grid[m - 1][j2]

    maxVal = int(-1e8)
    for dj1 in range(-1, 2):
        for dj2 in range(-1, 2):
            if j1 == j2:
                value = grid[i][j1]
            else:
                value = grid[i][j1] + grid[i][j2]

            value += cherryPickup2_dfs(grid, i + 1, j1 + dj1, j2 + dj2, m, n)
            maxVal = max(maxVal, value)

    return maxVal


# ---------------- Test (same as Java main) ----------------

grid = [
    [3, 1, 1],
    [2, 5, 1],
    [1, 5, 5],
    [2, 1, 1]
]

print(cherryPickup(grid))
print(cherryPickup2(grid))
